Adding preview of flip showing what cookbooks and versions will be applied by flipping to new env

jkim opened the pull request:
Adding in functionality to preview the node flip and see what cookbooks will have their differing changes applied.

jkim committed:
Adding support for previewing a node flip change before its made so you can see what cookbooks are different (and what versions) in the environment you are about to flip to

jonlives commented:
Jeff,

When I try running this locally using chef 10.24.0 with the --preview option, I'm getting this error:
$> knife node flip <snip>.com testing --preview -VV

```
/Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/chef-10.24.0/lib/chef/node/attribute.rb:445:in `method_missing': Attribute environment is not defined! (ArgumentError)
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/chef-10.24.0/lib/chef/node.rb:346:in `method_missing'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/knife-flip-0.1.7/lib/chef/knife/nodeflip.rb:122:in `get_source_environment'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/knife-flip-0.1.7/lib/chef/knife/nodeflip.rb:83:in `show_environmental_differences'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/knife-flip-0.1.7/lib/chef/knife/nodeflip.rb:35:in `run'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/chef-10.24.0/lib/chef/knife.rb:408:in `run_with_pretty_exceptions'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/chef-10.24.0/lib/chef/knife.rb:168:in `run'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/chef-10.24.0/lib/chef/application/knife.rb:123:in `run'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/gems/chef-10.24.0/bin/knife:25:in `<top (required)>'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/bin/knife:23:in `load'
    from /Users/jcowie/.rvm/gems/ruby-1.9.3-p0/bin/knife:23:in `<main>'
```

This error doesn't occur if I use chef 11.4.4. A lot of external flip users are still using Chef 10, so I'm reluctant to merge a change which breaks under Chef 10 - would be good to have this fixed.

Also when moving the above node into the (unconstrained) testing environment, I'd have expected to see a bunch of cookbooks listed which are now unconstrained, but it shows as no difference. 

I'd also like to make it a little clearer that --preview does not change the environment of the node, possible with an extra comment or something - it took me a while to realise that the env wasn't being changed.

jkim committed:
Putting in a git ignore, should be relatively safe patterns

Adding in colorize gem requirement for cleaner output

Fix to make preview work on both chef 10 and chef 11

Making the output clearer with regards to what the environment switch will do

jkim commented:
Jon,

I've pushed up some changes which should fix the Chef 10 issue, as well as make the output clearer with regards to what will change, and that --preview is a dry-run not-actually-flipping param

Lemme know your thoughts
